### PR TITLE
jobs-api

### PR DESCRIPTION
**Pull request by liaojason2softchef**

### CODE REVIEW

File: lambda-assets/jobs/list-job-execution-job/app.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  const request = new Request(event);
+  const response = new Response();
+  try {
+    let parameters: { [key: string]: any } = {};
+    if (request.has('nextToken')) {
+      parameters.nextToken = request.get('nextToken');
+    };
+    const iotClient = new IoTClient({});
+    const jobsExecutionJob = await iotClient.send(
+      new ListJobExecutionsForJobCommand({
```
Comment (Kerwin-Jhong) [inline lambda-assets/jobs/list-job-execution-job/app.ts:14]: status

File: lambda-assets/jobs/list-job-execution-thing/app.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  const request = new Request(event);
+  const response = new Response();
+  try {
+    let parameters: { [key: string]: any } = {};
+    if (request.has('nextToken')) {
+      parameters.nextToken = request.get('nextToken');
+    };
+    const iotClient = new IoTClient({});
+    const jobsExecutionThing = await iotClient.send(
+      new ListJobExecutionsForThingCommand({
```
Comment (Kerwin-Jhong) [inline lambda-assets/jobs/list-job-execution-thing/app.ts:14]: status

File: lambda-assets/jobs/associate-targets-with-job/app.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+export async function handler(event: { [key: string]: any }) {
+  const request = new Request(event);
+  const response = new Response();
+  try {
+    const iotClient = new IoTClient({});
+    await iotClient.send(
+      new AssociateTargetsWithJobCommand({
+        jobId: request.parameter('jobId'),
+        targets: request.input('targets'),
```
Comment (Kerwin-Jhong) [inline lambda-assets/jobs/associate-targets-with-job/app.ts:12]: Joi

File: lambda-assets/jobs/cancel-job/app.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import { Request, Response } from '@softchef/lambda-events';
+
+export async function handler(event: { [key: string]: any }) {
+  const request = new Request(event);
+  const response = new Response();
+  try {
+    const iotClient = new IoTClient({});
+    await iotClient.send(
+      new CancelJobCommand({
+        jobId: request.parameter('jobId'),
```
Comment (Kerwin-Jhong) [inline lambda-assets/jobs/cancel-job/app.ts:11]: comment
force

File: test/lambda/job.test.ts [hunk 5]
```diff
@@ -589,12 +745,266 @@ test('Get job execution with invalid thingName expect failure', async() => {
...
+
+test('List job executions for thing with nextToken success', async () => {
+  const iotClientMock = mockClient(IoTClient);
+  iotClientMock.on(ListJobExecutionsForThingCommand, {
+    thingName: expected.jobExecution.thingName,
+    nextToken: expected.listJobExecutionForThing.nextToken,
+  }).resolves({
+    executionSummaries: [{
+      jobId: expected.jobExecution.jobId,
+      jobExecutionSummary: {
```
Comment (Kerwin-Jhong) [inline test/lambda/job.test.ts:989]: from parameter

File: test/lambda/job.test.ts [hunk 6]
```diff
@@ -589,12 +745,266 @@ test('Get job execution with invalid thingName expect failure', async() => {
...
+
+test('List job executions for thing success', async () => {
+  const iotClientMock = mockClient(IoTClient);
+  iotClientMock.on(ListJobExecutionsForThingCommand, {
+    thingName: expected.jobExecution.thingName,
+    nextToken: expected.listJobExecutionForThing.nextToken,
+  }).resolves({
+    executionSummaries: [{
+      jobId: expectedJobExecution.jobId,
+      jobExecutionSummary: {
```
Comment (Kerwin-Jhong) [inline test/lambda/job.test.ts:958]: from parameter

File: test/lambda/job.test.ts [hunk 7]
```diff
@@ -589,12 +745,266 @@ test('Get job execution with invalid thingName expect failure', async() => {
...
-  // const body = JSON.parse(response.body);
+test('List job Executions For job with nextToken success', async () => {
+  const iotClientMock = mockClient(IoTClient);
+  iotClientMock.on(ListJobExecutionsForJobCommand, {
+    jobId: expected.jobExecution.jobId,
+    nextToken: expected.listJobExecutionForJob.nextToken,
+  }).resolves({
+    executionSummaries: [{
+      thingArn: expectedJobExecution.thingArn,
+      jobExecutionSummary: {
```
Comment (Kerwin-Jhong) [inline test/lambda/job.test.ts:927]: from parameter

File: test/lambda/job.test.ts [hunk 8]
```diff
@@ -589,12 +745,266 @@ test('Get job execution with invalid thingName expect failure', async() => {
...
+});
+
+test('List job executions for job success', async () => {
+  const iotClientMock = mockClient(IoTClient);
+  iotClientMock.on(ListJobExecutionsForJobCommand, {
+    jobId: expected.jobExecution.jobId,
+  }).resolves({
+    executionSummaries: [{
+      thingArn: expectedJobExecution.thingArn,
+      jobExecutionSummary: {
```
Comment (Kerwin-Jhong) [inline test/lambda/job.test.ts:897]: from parameter

File: test/lambda/job.test.ts [hunk 9]
```diff
@@ -341,13 +497,13 @@ test('Update job with invalid jobId expect failure', async() => {
   iotClientMock.restore();
 });
 
-test('Update job targets success', async() => {});
+test('Update job targets success', async () => { });
```
Comment (Kerwin-Jhong) [inline test/lambda/job.test.ts:500]: ?

File: test/lambda/job.test.ts [hunk 10]
```diff
@@ -341,13 +497,13 @@ test('Update job with invalid jobId expect failure', async() => {
...
 
-test('Update job targets with invalid inputs expect failure', async() => {});
+test('Update job targets with invalid inputs expect failure', async () => { });
```
Comment (Kerwin-Jhong) [inline test/lambda/job.test.ts:502]: ?

File: test/lambda/job.test.ts [hunk 11]
```diff
@@ -341,13 +497,13 @@ test('Update job with invalid jobId expect failure', async() => {
...
 
-test('Update job targets with invalid jobId expect failure', async() => {});
+test('Update job targets with invalid jobId expect failure', async () => { });
```
Comment (Kerwin-Jhong) [inline test/lambda/job.test.ts:504]: ?

**Review by Kerwin-Jhong — COMMENTED** [review]
請補上 API
src/job-api.ts

**Review by MinCheTsai — APPROVED** [review]

File: lambda-assets/jobs/associate-targets-with-job/app.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
+import { IoTClient, AssociateTargetsWithJobCommand } from '@aws-sdk/client-iot';
+import { Request, Response } from '@softchef/lambda-events';
+
+export async function handler(event: { [key: string]: any }) {
+  const request = new Request(event);
+  const response = new Response();
+  try {
+    const validated = request.validate(joi => {
+      return {
+        targets: joi.array().items(joi.string()).required(),
```
Comment (Kerwin-Jhong) [inline lambda-assets/jobs/associate-targets-with-job/app.ts:10]: joi.array().items(
  joi.string(),
).required(),

File: lambda-assets/jobs/get-job-document/app.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+        jobId: request.parameter('jobId'),
+      }),
+    );
+    return response.json(document);
+  } catch (error) {
+    if (error.Code === 'ResourceNotFoundException') {
+      return response.error(error, 404);
+    } else {
+      return response.error(error);
+    }
```
Comment (Kerwin-Jhong) [inline lambda-assets/jobs/get-job-document/app.ts:20]: ;

File: lambda-assets/jobs/cancel-job/app.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    );
+    return response.json({
+      canceled: true,
+    });
+  } catch (error) {
+    if (error.Code === 'ResourceNotFoundException') {
+      return response.error(error, 404);
+    } else {
+      return response.error(error);
+    }
```
Comment (Kerwin-Jhong) [inline lambda-assets/jobs/cancel-job/app.ts:24]: ;

File: lambda-assets/jobs/associate-targets-with-job/app.ts [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    );
+    return response.json({
+      associate: true,
+    });
+  } catch (error) {
+    if (error.Code === 'ResourceNotFoundException') {
+      return response.error(error, 404);
+    } else {
+      return response.error(error);
+    }
```
Comment (Kerwin-Jhong) [inline lambda-assets/jobs/associate-targets-with-job/app.ts:31]: ;

File: lambda-assets/jobs/cancel-job-execution/app.ts [hunk 16]
```diff
@@ -35,4 +35,4 @@ export async function handler(event: { [key: string]: any }) {
       return response.error(error);
     }
```
Comment (Kerwin-Jhong) [inline lambda-assets/jobs/cancel-job-execution/app.ts:36]: ;

File: lambda-assets/jobs/list-job-execution-job/app.ts [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+        ...parameters,
+      }),
+    );
+    return response.json(jobsExecutionJob);
+  } catch (error) {
+    if (error.Code === 'ResourceNotFoundException') {
+      return response.error(error, 404);
+    } else {
+      return response.error(error);
+    }
```
Comment (Kerwin-Jhong) [inline lambda-assets/jobs/list-job-execution-job/app.ts:26]: ;

File: lambda-assets/jobs/list-job-execution-thing/app.ts [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+        ...parameters,
+      }),
+    );
+    return response.json(jobsExecutionThing);
+  } catch (error) {
+    if (error.Code === 'ResourceNotFoundException') {
+      return response.error(error, 404);
+    } else {
+      return response.error(error);
+    }
```
Comment (Kerwin-Jhong) [inline lambda-assets/jobs/list-job-execution-thing/app.ts:26]: ;

File: test/lambda/job.test.ts [hunk 19]
```diff
@@ -589,12 +755,273 @@ test('Get job execution with invalid thingName expect failure', async() => {
...
+  const response = await ListJobExecutionsForThing.handler({
+    queryStringParameters: {
+      thingName: expected.jobExecution.thingName,
+      status: expected.jobExecution.status,
+      nextToken: expected.listJobExecutionForThing.nextToken,
+    },
+  });
+  const body = JSON.parse(response.body);
+  expect(Array.isArray(body.executionSummaries)).toBe(true);
+  expect(body.executionSummaries).toEqual(expected.listJobExecutionForThing.executionSummaries);
```
Comment (Kerwin-Jhong) [inline test/lambda/job.test.ts:1024]: expect(body).toEqual(expected.listJobExecutionForThing)

File: test/lambda/job.test.ts [hunk 20]
```diff
@@ -589,12 +755,273 @@ test('Get job execution with invalid thingName expect failure', async() => {
...
+  iotClientMock.restore();
+});
+
+test('List job executions for thing with nextToken success', async () => {
+  const iotClientMock = mockClient(IoTClient);
+  iotClientMock.on(ListJobExecutionsForThingCommand, {
+    thingName: expected.jobExecution.thingName,
+    nextToken: expected.listJobExecutionForThing.nextToken,
+  }).resolves({
+    executionSummaries: [{
```
Comment (Kerwin-Jhong) [inline test/lambda/job.test.ts:1003]: {
executionSummaries : expected.listJobExecutionForThing.executionSummaries,
nextToken : expected.listJobExecutionForThing. nextToken
}

**Review by MinCheTsai — APPROVED** [review]

**Pull request merged**